src/fpylll/tools/probabilistic_bkz_simulator.py

```diff
@@ -0,0 +1,215 @@
+# -*- coding: utf-8 -*-
+"""
+BKZ simulation algorithm as proposed in
+
+- Bai, S., & Stehlé, D. & Wen, W. (2018). Measuring, simulating and exploiting
+  the head concavity phenomenon in BKZ. In T. Peyrin, & S. Galbraith,
+  ASIACRYPT~2018 : Springer, Heidelberg.
+
+The code is based on the implementation of the simulation algorithm proposed by
+Chen and Nguyen in "BKZ 2.0: Better Lattice Security Estimates" available at
+https://github.com/fplll/fpylll/blob/master/src/fpylll/tools/bkz_simulator.py
+by Michael Walter and Martin R. Albrecht.
+
+.. moduleauthor:: Fernando Virdia <fernando.virdia.2016@rhul.ac.uk> (2020)
+
+"""
+from copy import copy
+from math import log, sqrt, lgamma, pi
+from collections import OrderedDict
+import random
+
+from fpylll.tools.quality import basis_quality
+from fpylll.tools.bkz_stats import pretty_dict
+from fpylll.fplll.bkz import BKZ
+from fpylll.fplll.integer_matrix import IntegerMatrix
+from fpylll.fplll.gso import MatGSO, GSO
+from fpylll import FPLLL
+
+
+rk = (
```

Comment:
I'd suggest to have this list only once in the library.

Reply:
will change